### PR TITLE
Add Cosmos3 Edge and Distilled checkpoints support

### PR DESCRIPTION
**Pull request by bastefaniak**
## Summary

This MR adds support for the upcoming Cosmos3 Edge and distilled models and
improves parity with the original implementation in `cosmos-framework`.

### Cosmos3 Edge

- Adds `Cosmos3EdgeVFMTransformer`, a Nemotron-based transformer architecture
  used by Cosmos3 Edge models.
- Selects the Edge transformer using the `backbone_type` field from
  `transformer/config.json`.

### Distilled models

- Distilled checkpoints differ from regular one only by scheduler used, detects them from `FlowMatchEulerDiscreteScheduler` in the scheduler configuration.
- Forces `guidance_scale=1.0` for distilled inference.

### Pipeline compatibility

- Registers `Cosmos3OmniPipeline` as an alias for
  `Cosmos3OmniDiffusersPipeline` to support the new checkpoint configuration.

### Parity with `cosmos-framework`

- Generates `TimestepEmbedder` frequencies in FP32.
- Loads JSON prompts and updates their metadata from request parameters.
- Replaces the previous `UniPCMultistepScheduler` approximation with the
  native `FlowUniPCMultistepScheduler` implementation.

> **Note:** The regression reported in
> [#4627](https://github.com/vllm-project/vllm-omni/pull/4627) was not caused by
> the scheduler change. It was caused by the video-generation default
> `flow_shift` being `1.0`; the default is now `10.0`.

**vLLM Version:**
0.25

**vLLM-Omni Commit:**
Top of tree of this MR

## Test Result

```
python3 -m pytest -q   tests/diffusion/models/cosmos3   tests/diffusion/cache/test_cache_dit.py   tests/config/test_pipeline_registry.py   tests/model_extras/test_model_extras.py
158 passed, 22 warnings in 1.74s
```
```
pytest -s -v -m "core_model and cpu"
4185 passed, 22 skipped, 2189 deselected, 72 warnings in 266.64s (0:04:26)
```
**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: recipes/cosmos3/Cosmos3-Edge.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+auto-resolves from `model_index.json`; pass
+`--model-class-name Cosmos3OmniDiffusersPipeline` to force it explicitly.
+
+#### Examples
+
+Example requests and reference outputs will be provided later.
+
+#### Notes
+
+- **Performance:** TBD. Checkpoint is not yet released
```
Comment (hsliuustc0106) [inline recipes/cosmos3/Cosmos3-Edge.md]: next PR for updates?

File: vllm_omni/diffusion/cache/cache_dit_backend.py [hunk 2]
```diff
@@ -1016,6 +1016,7 @@ def enable_cache_for_krea2(pipeline: Any, cache_config: Any) -> RefreshCacheCont
         "Wan22VACEPipeline": enable_cache_for_wan22,
         "Wan22S2VPipeline": enable_cache_for_wan22_s2v,
         "Cosmos3OmniDiffusersPipeline": enable_cache_for_cosmos3,
+        "Cosmos3OmniPipeline": enable_cache_for_cosmos3,
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/cache/cache_dit_backend.py:1019]: this one is for edge? if so can we rename the pipelines?

File: docs/models/supported_models.md [hunk 3]
```diff
@@ -37,7 +37,7 @@ th {
 | `Krea2Pipeline` | Krea 2 (Raw + Turbo) | `krea/Krea-2-Raw`, `krea/Krea-2-Turbo` | ✅︎ | | | |
 | `WanPipeline` | Wan2.1-T2V, Wan2.2-T2V, Wan2.2-TI2V | `Wan-AI/Wan2.1-T2V-1.3B-Diffusers`, `Wan-AI/Wan2.1-T2V-14B-Diffusers`, `Wan-AI/Wan2.2-T2V-A14B-Diffusers`, `Wan-AI/Wan2.2-TI2V-5B-Diffusers` | ✅︎ | ✅︎ | ✅︎ | ✅︎ |
 | `WanImageToVideoPipeline` | Wan2.2-I2V | `Wan-AI/Wan2.2-I2V-A14B-Diffusers` | ✅︎ | ✅︎ | ✅︎ | ✅︎ |
-| `Cosmos3OmniDiffusersPipeline` | Cosmos3 T2I, T2V, I2V, V2V, T2V with sound, action policy | `nvidia/Cosmos3-Nano`, `nvidia/Cosmos3-Super` | ✅︎ | ✅︎ | ✅︎ | ✅︎ |
+| `Cosmos3OmniDiffusersPipeline` | Cosmos3 T2I, T2V, I2V, V2V, T2V with sound, action policy; Cosmos3 Edge T2I/T2V/I2V/action | `nvidia/Cosmos3-Nano`, `nvidia/Cosmos3-Super`, `nvidia/Cosmos3-Edge` | ✅︎ | ✅︎ | ✅︎ | ✅︎ |
```
Comment (hsliuustc0106) [inline docs/models/supported_models.md]: This marks Cosmos3 Edge as fully supported, but the PR does not include Edge run evidence yet. Please add runnable Edge online/offline commands, sample output, latency, and peak VRAM, or keep the support row scoped until that evidence is available.

File: recipes/cosmos3/Cosmos3-Edge.md (hunk 1 above)
Comment (hsliuustc0106) [inline recipes/cosmos3/Cosmos3-Edge.md]: Need actual Edge latency and peak VRAM here before merge, with hardware, resolution, step count, and the command used to measure. `TBD` makes the new model support unverifiable.

File: docs/models/supported_models.md (hunk 3 above)
Comment (bastefaniak) [reply docs/models/supported_models.md]: Would it be fine, if I removed docs changes and created followup PR when checkpoint is oficially released?

File: vllm_omni/diffusion/cache/cache_dit_backend.py (hunk 2 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/cache/cache_dit_backend.py:1019]: The new name is required by `diffusers`. I wanted to have both names to remain backward compatible with previously downloaded checkpoints

File: recipes/cosmos3/Cosmos3-Edge.md (hunk 1 above)
Comment (bastefaniak) [reply recipes/cosmos3/Cosmos3-Edge.md]: Would it be fine, if I removed docs changes and created followup PR when checkpoint is oficially released?

Comment (bastefaniak) [reply recipes/cosmos3/Cosmos3-Edge.md]: Would it be fine, if I removed docs changes and created followup PR when checkpoint is oficially released?

File: docs/models/supported_models.md (hunk 3 above)
Comment (Gaohan123) [reply docs/models/supported_models.md]: I think so

File: vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py [hunk 4]
```diff
@@ -407,7 +406,7 @@ def __init__(
         self.hidden_size = hidden_size
 
         half = frequency_embedding_size // 2
-        freqs = torch.exp(-math.log(max_period) * torch.arange(start=0, end=half, dtype=target_dtype) / half)
+        freqs = torch.exp(-math.log(max_period) * torch.arange(start=0, end=half, dtype=torch.float32) / half)
```
Comment (Gaohan123) [inline vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py:409]: Why do we change it? It seems to affect the accuracy

Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py:409]: I've changed it for better [parity](https://github.com/NVIDIA/cosmos-framework/blob/main/cosmos_framework/model/generator/mot/modeling_utils.py#L63) with `cosmos-framework` that contains source of truth for code. I verified on internal benchmarks, that quality remains very similar, but it's easier to debug when we are as close to original code as possible

File: recipes/cosmos3/Cosmos3-Edge.md (hunk 1 above)
Comment (bastefaniak) [reply recipes/cosmos3/Cosmos3-Edge.md]: Removed Cosmos3-Edge from docs, will add in followup PR when checkpoint is released

File: docs/models/supported_models.md (hunk 3 above)
Comment (bastefaniak) [reply docs/models/supported_models.md]: Removed Cosmos3-Edge from docs, will add in followup PR when checkpoint is released

File: recipes/cosmos3/Cosmos3-Edge.md (hunk 1 above)
Comment (bastefaniak) [reply recipes/cosmos3/Cosmos3-Edge.md]: Removed Cosmos3-Edge from docs, will add in followup PR when checkpoint is released

File: vllm_omni/diffusion/models/cosmos3/transformer_cosmos3_edge.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+        hidden_states = residual + attn_out
+
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        hidden_states = residual + self.mlp(hidden_states)
+
+        return hidden_states, k, v
+
+
+class Cosmos3EdgeLanguageModel(nn.Module):
```
Comment (david6666666) [inline vllm_omni/diffusion/models/cosmos3/transformer_cosmos3_edge.py:241]: `Cosmos3LanguageModel` declares `_layerwise_offload_blocks_attrs = ["layers"]`; this sibling does not, so the Edge UND tower is skipped by layerwise CPU offload. Please add the same attribute and cover it in the offload contract test.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 6]
```diff
@@ -764,27 +839,43 @@ def __init__(
...
-        if od_config.flow_shift is not None:
-            self.scheduler = UniPCMultistepScheduler.from_config(
-                self.scheduler.config,
-                flow_shift=od_config.flow_shift,
+
+        scheduler_class_name = scheduler_config.get("_class_name")
+
+        self.is_distilled_model = False
+        if scheduler_class_name == COSMOS3_DISTILLED_CHECKPOINT_SCHEDULER_CLASS:
+            self.scheduler = FlowMatchEulerDiscreteScheduler.from_config(scheduler_config)
```
Comment (david6666666) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: `FlowMatchEulerDiscreteScheduler` defaults to deterministic Euler, while Cosmos3 distilled sampling re-noises at every fixed step. Please force or validate `stochastic_sampling=True` here and cover an actual scheduler step.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 7]
```diff
@@ -1484,38 +1582,42 @@ def _is_t2i_request(req: DiffusionRequestBatch) -> bool:
...
+
+    def _set_flow_shift(self, target_shift: float) -> None:
+        """Select the shift applied by FlowUniPC when timesteps are built."""
+        self._current_flow_shift = float(target_shift)
+
+    def _seed_global_rng(self, seed: int | None) -> None:
+        if seed is None:
             return
+        seed = int(seed)
+        torch.manual_seed(seed)
```
Comment (david6666666) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: This mutates process-global RNG state for every request, and overlapping requests can interfere. Please thread the request-local generator into `scheduler.step()` instead.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 8]
```diff
@@ -1484,38 +1582,42 @@ def _is_t2i_request(req: DiffusionRequestBatch) -> bool:
...
+        if sp.seed is not None:
+            return int(sp.seed)
+        if isinstance(generator, torch.Generator):
+            return int(generator.initial_seed())
+        return int(default)
+
+    def _resolve_guidance_scale(self, sp: OmniDiffusionSamplingParams, default: float) -> float:
+        if self.is_distilled_model:
+            return 1.0
+        if sp.guidance_scale_provided:
```
Comment (david6666666) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:1695]: The new `0.0` case is not reachable through a real request: `OmniDiffusionRequest.__post_init__` treats zero as omitted and overwrites it with `1.0`. Please fix the sentinel handling or drop that contract.

File: vllm_omni/diffusion/models/cosmos3/transformer_cosmos3_edge.py (hunk 5 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/transformer_cosmos3_edge.py:241]: Added

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 6 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: We store that in checkpoint, but added forcing, tests and validation of other parameters

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 7 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: Fixed, note that `FlowMatchEulerDiscreteScheduler` doesn't respect generator in `diffusers==0.38.0`, it's fixed in `diffusers==0.39.0` (but we should be fine for now).

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 8 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:1695]: Is this comment about unittests? If so updated them

**Review by ywang96 — APPROVED** [review]
